### PR TITLE
Fix: Make Archipelago flow truly recoverable

### LINKED ISSUE #2686: InvalidOperationException: Connection is not established blocks the loading screen
Over the "Loading screen block issue", we found this one

https://github.com/decentraland/unity-explorer/issues/2297#issuecomment-2447541914
Player Log: https://github.com/user-attachments/files/17575260/Player.log

```
InvalidOperationException: Connection is not established
  at DCL.Multiplayer.Connections.Archipelago.SignFlow.LiveConnectionArchipelagoSignFlow.StartListeningForConnectionStringAsync (System.Action`1[T] onNewConnectionString, System.Threading.CancellationToken token) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTaskVoid`1[TStateMachine].Run () [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1[TResult].TrySetResult (TResult result) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`2[TStateMachine,T].SetResult (T result) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTaskMethodBuilder`1[T].SetException (System.Exception exception) [0x00000] in <00000000000000000000000000000000>:0 
  at DCL.Multiplayer.Connections.Archipelago.LiveConnections.LogArchipelagoLiveConnection.ReceiveAsync (System.Threading.CancellationToken token) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`2[TStateMachine,T].Run () [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.UniTaskCompletionSourceCore`1[TResult].TrySetResult (TResult result) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`2[TStateMachine,T].SetResult (T result) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTaskMethodBuilder`1[T].SetException (System.Exception exception) [0x00000] in <00000000000000000000000000000000>:0 
  at DCL.Multiplayer.Connections.Archipelago.LiveConnections.WebSocketArchipelagoLiveConnection.ReceiveAsync (System.Threading.CancellationToken token) [0x00000] in <00000000000000000000000000000000>:0 
  at Cysharp.Threading.Tasks.CompilerServices.AsyncUniTask`2[TStateMachine,T].Run () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.UnitySynchronizationContext+WorkRequest.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.UnitySynchronizationContext.Exec () [0x00000] in <00000000000000000000000000000000>:0 
Rethrow as Exception: Cannot listen for connection string
```

### PR DESCRIPTION
**Pull request by mikhail-dcl**
The previous flow:
- Didn't really use the auto-reconnection flow, that flow itself was a bit flawed
- `StartListeningForConnectionStringAsync` could be interrupted and never re-launched again, `catch` was outside of the loop
- There was a mix of `Result` and `Exceptions`: see [doc](https://github.com/decentraland/unity-explorer/wiki/Architecture-Overview#exceptions-free-async-flow)

Solution:
* Make the auto-reconnection flow be resistent to exceptions and interrupt only on cancellation. On Cancellation the flow will be re-launched from the connective room itself 
* Align and secure with `Result`/`LightResult` to exclude exceptions from the flow
* Enforce AutoReconnection
* Cover by unit tests

Technical thoughts, left for later
* `Archipelago` flow can be simplified: currently, it's broken down into 3 independent layers, each of them is represented by a series of decorators. It feels unjustifiedly complicated and quite hard to follow.
* There are multiple ways to achieve an exception-free flow: `Result`, `LightResult`, and `EnumResult<TErrorEnum>`. There could be a more generalized and clarified way of achieving essentially the same

Fixes #2686, #2447

## How to test the changes?

- Test the whole flow of Islands change: jumping between coordinates in Genesis City, and worlds
- Try to turn off your internet connection while `Island`:
   - Is being connected to (not connected yet)
   - Already connected
   - During teleportation (?)
- `Island` should recover when `Internet` is back in every case

### CODE REVIEW

**Comment by NickKhalow** [comment]
> Archipelago flow can be simplified: currently, it's broken down into 3 independent layers, each of them is represented by a series of decorators. It feels unjustifiedly complicated and quite hard to follow.

How would you simplify the flow and what makes it hard to follow in your opinion?

**Review by AnsisMalins — COMMENTED** [review]
I find no faults, but I feel unqualified to approve this.

**Review by m3taphysics — APPROVED** [review]
